### PR TITLE
backend-plugin-api: remove plugin and module options

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Getting rid of options, we found them to encourage suboptimal design and don't work well with declarative integration of plugins. Both plugins and modules should now instead either rely on static configuration or extension points for customization.

The build for this one is gonna be broken for a while until some more PRs are merged.

Turns out scaffolder options were still there, so this PR also removes the scaffolder plugin options and creates a few new extension points.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by vinzscam — APPROVED** [review]

File: plugins/scaffolder-backend/src/ScaffolderPlugin.ts [hunk 1]
```diff
@@ -23,120 +23,104 @@ import { ScmIntegrations } from '@backstage/integration';
...
+      },
+    });
 
-  addActions(...actions: TemplateAction<any>[]): void {
-    this.#actions.push(...actions);
-  }
+    const additionalTemplateFilters: Record<string, TemplateFilter> = {};
+    const additionalTemplateGlobals: Record<string, TemplateGlobal> = {};
+    env.registerExtensionPoint(scaffolderTemplatingExtensionPoint, {
+      addTemplateFilters(newFilters) {
```
Comment (freben) [inline plugins/scaffolder-backend/src/ScaffolderPlugin.ts:64]: just noting that we sometimes tend to prefer add-single-thing-at-a-time interfaces, but this is fine

File: plugins/scaffolder-backend/src/ScaffolderPlugin.ts [hunk 2]
```diff
@@ -23,120 +23,104 @@ import { ScmIntegrations } from '@backstage/integration';
...
- *
- * @alpha
- */
-export const scaffolderPlugin = createBackendPlugin(
-  (options?: ScaffolderPluginOptions) => ({
-    pluginId: 'scaffolder',
-    register(env) {
-      const actionsExtensions = new ScaffolderActionsExtensionPointImpl();
+        const actions = [
+          ...addedActions,
```
Comment (freben) [inline plugins/scaffolder-backend/src/ScaffolderPlugin.ts:94]: no overriding here? as in, should these be spread in last or first, or doesn't it matter

**Review by freben — APPROVED** [review]

File: plugins/scaffolder-backend/src/ScaffolderPlugin.ts (hunk 2 above)
Comment (Rugvip) [reply plugins/scaffolder-backend/src/ScaffolderPlugin.ts:94]: Kept it as it was, but was also a bit curious about what makes sense here in practice. Thinking we could tweak it in the future if needed. The main thing I was worried about was not being able to disable the built-in actions, it does feel a bit awkward to add that as an extension point though. Maybe a `scaffolder.disableBuildingActions` config is the way to go?

File: plugins/scaffolder-backend/src/ScaffolderPlugin.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/scaffolder-backend/src/ScaffolderPlugin.ts:64]: Yeah was considering redesigning this a bit but in the end figured this was alright too

File: plugins/scaffolder-backend/src/ScaffolderPlugin.ts (hunk 2 above)
Comment (freben) [reply plugins/scaffolder-backend/src/ScaffolderPlugin.ts:94]: As with any "collection like" thing I think it's the most robust if we always go for complete replacement but with facilities for easily reaching the builtins. Same as with table columns etc, lots of prior art there

**Pull request merged**